crates/sui-core/src/epoch/reconfiguration.rs

```diff
@@ -0,0 +1,135 @@
+                            .to_owned(),
+                }
+            );
+            // drop checkpoints lock
+        }
+
+        self.state.halted.store(true, Ordering::SeqCst);
+        let instant = Instant::now();
+        while !self.state.batch_notifier.ticket_drained() {
+            tokio::time::sleep(Duration::from_millis(10)).await;
+            fp_ensure!(
+                instant.elapsed() <= MAX_START_EPOCH_WAIT_SECONDS,
+                SuiError::InconsistentEpochState {
+                    error:
+                        "Waiting for batch_notifier ticket to drain timed out in start_epoch_change"
+                            .to_owned(),
+                }
+            );
+        }
+        Ok(())
+    }
+
+    pub async fn finish_epoch_change(&self) -> SuiResult {
+        fp_ensure!(
+            self.state.halted.load(Ordering::SeqCst),
+            SuiError::InconsistentEpochState {
+                error: "finish_epoch_change called when validator is not halted".to_owned(),
+            }
+        );
+        {
+            let checkpoints = self.state.checkpoints.as_ref().unwrap().lock();
+            let next_cp = checkpoints.get_locals().next_checkpoint;
+            fp_ensure!(
+                Self::is_last_checkpoint_epoch(next_cp),
+                SuiError::InconsistentEpochState {
+                    error: "finish_epoch_change called at the wrong checkpoint".to_owned(),
+                }
+            );
+            fp_ensure!(
+                checkpoints.lowest_unprocessed_checkpoint() == next_cp,
+                SuiError::InconsistentEpochState {
+                    error:
+                        "finish_epoch_change called when there are still unprocessed transactions"
+                            .to_owned(),
+                }
+            );
+            if checkpoints.extra_transactions.iter().next().is_some() {
+                // TODO: Revert any tx that's executed but not in the checkpoint.
+            }
+            // drop checkpoints lock
+        }
+
+        let sui_system_state = self.state.get_sui_system_state_object().await?;
+        let next_epoch = sui_system_state.epoch + 1;
+        let next_epoch_validators = &sui_system_state.validators.next_epoch_validators;
+        let votes = next_epoch_validators
+            .iter()
+            .map(|metadata| {
+                (
+                    PublicKeyBytes::try_from(metadata.pubkey_bytes.as_ref())
+                        .expect("Validity of public key bytes should be verified on-chain"),
+                    metadata.next_epoch_stake,
+                )
+            })
+            .collect();
+        let new_committee = Committee::new(next_epoch, votes);
+        self.state.insert_new_epoch_info(&new_committee)?;
+        self.state.checkpoints.as_ref().unwrap().lock().committee = new_committee;
+        // TODO: Update all committee in all components, potentially restart some authority clients.
```

Comment:
Honesly, here I would bring down all active processes incl. narwhal, and start fresh instances with the new committee.